### PR TITLE
Productions 3 and 4

### PR DESCRIPTION
**Pull request by a-float**

### CODE REVIEW

File: src/__test__/test_production_3.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
+import unittest
+import itertools as it
+
+from graph import Graph
+from model import Node, NodeAttrs, EdgeAttrs, Edge
+
+from production import P3
+
+class TestProduction2(unittest.TestCase):
```
Comment (kkafar) [inline src/__test__/test_production_3.py:9]:
```suggestion
class TestProduction3(unittest.TestCase):
```

Comment (kkafar) [reply src/__test__/test_production_3.py:9]: Same for production 4

File: src/__test__/test_production_3.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        graph.add_q_hyperedge((node_0, node_1, node_2, node_3), EdgeAttrs('q', True))
+
+        for node_1, node_2 in it.pairwise(nodes + [node_0]):
+            graph.add_edge(Edge(node_1.handle, node_2.handle, EdgeAttrs(kind='e', flag=False)))
+
+        p3 = P3()
+        mapping_gen = graph.generate_subgraphs_isomorphic_with(p3.get_lhs())
+        self.assertTrue(any(p3.is_mapping_feasible(graph, mapping) for mapping in mapping_gen))
+
+    def test_production_can_be_applied_if_hanging_nodes_in_different_places(self):
```
Comment (kkafar) [inline src/__test__/test_production_3.py:31]: These tests are effectively useless, as `is_mapping_feasible` returns `True` by default unless overridden

We should rather test whether graph structure after applying the production looks like expected.

I believe you just followed other unittests, but they are also useless after #17 was merged, thus either remove them completely or let's made them actually do something.

File: src/__test__/test_production_3.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+        graph.add_node_collection(nodes)
+        graph.add_q_hyperedge((node_0, node_1, node_2, node_3), EdgeAttrs('q', True))
+
+        for node_1, node_2 in it.pairwise(nodes + [node_0]):
+            graph.add_edge(Edge(node_1.handle, node_2.handle, EdgeAttrs(kind='e', flag=False)))
+
+        p3 = P3()
+        mapping_gen = graph.generate_subgraphs_isomorphic_with(p3.get_lhs())
+        self.assertTrue(any(p3.is_mapping_feasible(graph, mapping) for mapping in mapping_gen))
```
Comment (kkafar) [inline src/__test__/test_production_3.py:50]: Yeah, same here

File: src/__test__/test_production_4.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import unittest
+import itertools as it
+
+from graph import Graph
+from model import Node, NodeAttrs, EdgeAttrs, Edge
+
+from production import P4
+
+class TestProduction2(unittest.TestCase):
```
Comment (kkafar) [inline src/__test__/test_production_4.py:9]:
```suggestion
class TestProduction4(unittest.TestCase):
```

**Review by kkafar — COMMENTED** [review]
I'm not testing whether productions work as expected, just looking at the code.

I left a remark to unit tests, especially these with `assertTrue`.
Negative tests `assertFalse` pass because there is most likely no mapping returned and that is what we should check explicitly. 

Besides that looks good

File: src/__test__/test_production_4.py (hunk 4 above)
Comment (a-float) [reply src/__test__/test_production_4.py]: I added a commit fixing test class names and changing assertions to what we used during class

**Review by kkafar — APPROVED** [review]
🟢

**Pull request merged**